L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,462 @@
+5. users could extend `CertificateProvider` to define their own provider implementations
+
+As a reference, here is how `CertificateProvider` might be defined in C++:
+```cpp
+namespace grpc {
+
+template <class Subclass>
+class CertificateProvider {
+ public:
+  CertificateProvider()
+      : distributor_(grpc_tls_certificate_distributor_create()) {
+    base.get_distributor = GetDistributor;
+    base.destroy = Destroy;
+    grpc_tls_certificate_distributor_set_watch_status_callback(
+        OnWatchStatusChangedCallback, this);
+  }
+
+  virtual ~CertificateProvider() {
+    grpc_tls_certificate_distributor_destroy(distributor_);
+  }
+
+  grpc_tls_certificate_provider* c_provider() const {
+    return grpc_tls_certificate_provider_external_create(&base_);
+  }
+
+  virtual void OnWatchStatusChanged(
+      const std::string& cert_name, bool watching_root,
+      bool watching_identity) = 0;
+
+  void SetRootCert(const std::string& cert_name,
+                   const std::string& contents) {
+    grpc_tls_certificate_distributor_set_root_certs(
+        distributor_, cert_name.c_str(), contents.c_str());
+  }
+
+  // ...similar functions for SetCertKeyPairs() and SetKeyMaterials()...
+
+ private:
+  static grpc_tls_certificate_distributor* GetDistributor(
+      grpc_tls_certificate_provider_external* external_provider) {
+    auto* provider = static_cast<Subclass*>(external_provider);
+    return provider->distributor_;
+  }
+
+  static void Destroy(
+      grpc_tls_certificate_provider_external* external_provider) {
+    auto* provider = static_cast<Subclass*>(external_provider);
+    delete provider;
+  }
+
+  static void OnWatchStatusChangedCallback(
+      void* arg, const std::string& cert_name, bool watching_root,
+      bool watching_identity) {
+    auto* provider = static_cast<Subclass*>(arg);
+    provider->OnWatchStatusChanged(cert_name, watching_root,
+                                   watching_identity);
+  }
+
+  // First data element to simulate inheritance.
+  grpc_tls_certificate_provider_external base_;
+
+  grpc_tls_certificate_distributor* distributor_;
+};
+
+}  // namespace grpc
+```
+and an example of custom provider implementation in C++:
+```cpp
+class MyCertificateProvider: public grpc::CertificateProvider<MyCertificateProvider> {
+ public:
+  void OnWatchStatusChanged(
+      const std::string& cert_name, bool watching_root,
+      bool watching_identity) override {
+    if (!watching_root && !watching_identity) {
+      certs_watching_.erase(cert_name);
+    } else {
+      certs_watching_[cert_name] = {watching_root, watching_identity};
+      // ...start or stop watching as needed...
+    }
+  }
+
+ private:
+  struct CertsWatching {
+    bool watching_root = false;
+    bool watching_identity = false;
+  };
+  std::map<std::string /*cert_name*/, CertsWatching> certs_watching_;
+};
+```
+### Custom Authorization
```

Comment:
@jiangtaoli2016 What's the plan for supporting async authorization check? The current server authorization check supports both sync and async, which is pretty complicated due to async support and goes against the goal of simple C-core API. The API's in this section can be simplified a lot if we decide to only support sync authorization check.

Reply:
I think async authz check has its values, especially considering the fact that a check can take a long time and we don't want to block the main thread. 
I was thinking for a long time if we can make `grpc_tls_authorization_check_arg` a class instead of a struct? Currently it contains a lot of information that the caller(from the wrap languages) doesn't need to know. That may simplify the logic a little bit, but I am not sure how long the refactoring would take...

Reply:
If we decide to support async, we first need to flush out some details on how to wrap the C-core API's related to server authorization check (e.g., `grpc_tls_authorization_check_arg` and others).

Reply:
OK. How about me writing a one pager briefly describe the necessary changes this would take?

Reply:
Yeah, but it also needs to be signed off by TL's of each wrapped language, which may take longer than we expect. Do we have an immediate use case for async authorization check @jiangtaoli2016? If we do not, how about we start with a simple sync version in this gRFC and introduce an async version in another gRFC when its design is mature?

Reply:
We  need to support async authz check, e.g., if customer needs to validate the cert chain with a remote server or fetching server authz ACLs remotely. gRPC core cannot be blocked. This is a requirement from @markdroth .

Yihua: we can re-use your previous server authorization check cb API, which was reviewed by Mark. Any concern with it?

Reply:
At the time I worked on async server authz check, I did not consider much about whether it can be easily wrapped by languages other than C++, while gRFC aims for the API simplicity for ALL wrapped languages. I am not sure if `struct` and function pointer fields in `grpc_tls_authorization_check_arg` can be easily wrapped.